### DISCUSSION TITLE
Module Versioning

### DISCUSSION THREAD

**Discussion by hishamco** [OrchardCMS/OrchardCore#11169] [General]
I just notice that on each release we bumb the versions for each NuGet package, while the OC modules are separated I think its good to track the changes that happen in every module dependently. Me as I user when I see OC.Html `1.1.0` for instance I'm expecting there's a new enhancements in that module, while the fact bump happen for each package regardless it's changed or not

I'd like to hear your thoughts about that

**Comment by Skrypt** [comment]
I think it is easier to update every module version because even though there is no change in one module; we update its version so that we know that all these modules work together with that same version.

```C#
[assembly: AssemblyInformationalVersion("1.0.0")]
```

The AssemblyInformationalVersion could be used independently to inform of a module version change. Though, this would be informational only. And at the same time, I'm not sure I like the idea because sometimes we update the .NET Core SDK version to build these modules on and that means we would need to update manually these module versions when that happens.

**Reply by hishamco** [reply]
Another thing I forgot to mention, when I have OC.Html `2.0.0` should it works for OC `2.0.0` only or all the preceeding versions as well

**Reply by Skrypt** [reply]
You can assume that 2.0.0 works with 2.0.0. But everything under that you can't assume it will work because it could have been built with .NET Core 3.1 for example ...

Semver : https://semver.org/

**Reply by hishamco** [reply]
I don't think we have notion of target specific OC version. For instance if I develop a module that target .NET Core 5.0, then assume there are two versions of OC target that version say: `1.2.0` and `1.3.0`. Now when I build a module for `1.2.0`, it should run for both OC version. Right?

**Reply by Skrypt** [reply]
1.2.0 was using multi-target so yes it should work but it is not right to assume that we will always multi-target. And of course, a module built for 1.2.0 should generally work for 1.3.0 but you can't assume it will work if there are deprecated methods in 6.0 that are used in your module that targets 5.0. It is not good to assume things will just work because we used to multi-target in 1.2.0. So, yes, we have a notion of a specific OC version. That's why we also have 1.3.0-preview to test things.

**Comment by hishamco** [comment]
Forget about deprecated method, the point that may confuse some people there is no notion of OC version in the modules, so the module consume knows if the module will work or not. May we don't have this because we don't rely on adding module dynamically at runtime like some other plugin systems

**Reply by Skrypt** [reply]
Normally when I update Orchard Core version in my Nuget package solutions I use a common .props file to update everything at the same time.

Here is what I use : 

```XML
<Project>

  <!-- Default TFMs to use when building and packaging -->
  <PropertyGroup>

    <!-- TFM used when coding in Visual Studio and when creating templates -->
    <DefaultTargetFramework>net6.0</DefaultTargetFramework>
    <OrchardCoreVersion>1.3.0-preview-16692</OrchardCoreVersion>

    <!-- TFMs used to build the abstractions and modules-->
    <CommonTargetFrameworks Condition="'$(CommonTargetFrameworks)' == ''">net6.0</CommonTargetFrameworks>
  </PropertyGroup>

  <!-- Detect if the solution is opened in VS to limit the TFMs that are analyzed by Roslyn for performance reasons -->
  <PropertyGroup Condition="'$(BuildingInsideVisualStudio)' == 'true' or '$(BuildingByReSharper)' == 'true'">
    <CommonTargetFrameworks>$(DefaultTargetFramework)</CommonTargetFrameworks>
  </PropertyGroup>

  <!-- These versions are used for the NuGet packages that are dependent on the current TFM -->
  <PropertyGroup Condition="'$(TargetFramework)' == 'net6.0'">
    <AspNetCorePackagesVersion>6.0.0</AspNetCorePackagesVersion>
    <MicrosoftExtensionsPackagesVersion>6.0.0</MicrosoftExtensionsPackagesVersion>
  </PropertyGroup>

  <!-- 'Microsoft.AspNetCore' packages that are not included in the ASP.NET Core shared framework -->
  <ItemGroup>
    <PackageManagement Include="Microsoft.AspNetCore.Authentication.AzureAD.UI" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Authentication.Facebook" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Authentication.Google" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Authentication.MicrosoftAccount" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Authentication.OpenIdConnect" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Authentication.Twitter" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.DataProtection.StackExchangeRedis" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Mvc.Testing" Version="$(AspNetCorePackagesVersion)" />
    <PackageManagement Include="Microsoft.AspNetCore.Owin" Version="$(AspNetCorePackagesVersion)" />
  </ItemGroup>

  <!-- 'Microsoft.Extensions' packages that are not included in the ASP.NET Core shared framework -->
  <ItemGroup>
    <PackageManagement Include="Microsoft.Extensions.Caching.StackExchangeRedis" Version="$(MicrosoftExtensionsPackagesVersion)" />
    <PackageManagement Include="Microsoft.Extensions.Http.Polly" Version="$(MicrosoftExtensionsPackagesVersion)" />
  </ItemGroup>

</Project>
```

So, here, as you can see I can update Orchard Core version simply in that file so that it tries to update all the Nuget packages at once. 

```xml
<OrchardCoreVersion>1.3.0-preview-16692</OrchardCoreVersion>
```

So, this is how I update Orchard Core version. So, there is a notion of version when using Nuget packages.

But, what you are suggesting is to package each module independently. This would mean that when there is an update made in a module we would need to update its version based on SemVer and then we would end up having unmatched modules version with Orchard Core.

Of course, this could be done but then we would need to separate Orchard Core framework project from the CMS modules to do so. And when we build modules with the CI we would need to have the version of the module set in the properties of each module themselves.

This would probably take a lot of effort for really few advantages over what we currently have. And we probably have a lot more work to do first about separating the actual framework from the CMS.

**Reply by Skrypt** [reply]
Then, we could use the same principle eventually by using a .props file but with a `<OrchardCoreFrameworkVersion>` ...
and a `<OrchardCoreCMSVersion>` ... but then that would mean that we would still need to have a common version for the modules still. So, to be honest, having the modules use different versions than the actual CMS here would still be an issue. You're supposed to be able still to reference the OC module version you want individually but at some point, you will end up with other modules' version dependencies mismatching errors.

**Reply by hishamco** [reply]
Thanks @Skrypt, seems you got my point. I know we have a lot of work to be done, regarding the separation from API level could be could in `2.0.0` or `3.0.0` because it's useful for OC community when they want to contrbiute

**Reply by hishamco** [reply]
At least we can separate the APIs using solutions, like what I did in #10436
